src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,193 @@
+        tail = newNode;
+        newNode.prev = tail.prev;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0 && size == 0) {
+            linkEmptyList(value);
+            return;
+        }
+        if (index == 0) {
+            linkHead(value);
+            return;
+        }
+        if (index == size) {
+            linkTail(value);
+            return;
+        }
+        checkIndex(index);
+        Node<T> current = getNnode(index);
+        Node<T> newNode = new Node<>(current.prev,value,current);
+        current.prev.next = newNode;
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.size() == 0) {
+            return;
+        }
+        for (T element:list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNnode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        final Node<T> current = getNnode(index);
+        final T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        return unlink(getNnode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (x.value == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> x = head; x != null; x = x.next) {
+                if (object.equals(x.value)) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void linkEmptyList(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        head = newNode;
+        tail = newNode;
+        size++;
+    }
+
+    private void linkHead(T value) {
+        final Node<T> f = head;
+        final Node<T> newNode = new Node<>(null, value, f);
+        head = newNode;
+        f.prev = newNode;
+        size++;
+    }
+
+    private void linkTail(T value) {
+        Node<T> l = tail;
+        Node<T> newNode = new Node<>(l, value,null);
+        tail = newNode;
+        l.next = newNode;
+        size++;
+    }
+
+    private Node<T> getNnode(int index) {
+        if (index < size >> 1) {
+            Node<T> x = head;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
```

Comment:
<img width="468" alt="image" src="https://github.com/mate-academy/jv-linked-list/assets/72043323/a3f1c730-5102-4973-9dcb-a2b3a332e31c">

we can avoid else block